### ISSUE TITLE
ec2_eni unable to create new ENIs with secondary private IPs

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29360]
From @KpaBap on 2016-09-18T23:05:31Z

##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

ec2_eni
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel fa5f8a7543) last updated 2016/09/18 02:10:05 (GMT -700)
  lib/ansible/modules/core: (detached HEAD 488f082761) last updated 2016/09/18 02:10:48 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 24da3602c6) last updated 2016/09/18 02:10:57 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible']
```
##### CONFIGURATION

Default 
##### OS / ENVIRONMENT

N\A
##### SUMMARY

I'm trying to create a new ENI in AWS with secondary private IPs.
Instead of creating new ENIs, Ansible just fails because it seems to be finding an existing ENI and refusing to change it.
##### STEPS TO REPRODUCE

```
- name: Create ENI for eth0 on Edge instances
  ec2_eni:
    state: present
    secondary_private_ip_address_count: 3
    subnet_id: subnet-72870d2a
    security_groups:
      - sg-5f4c6f27 
```

##### EXPECTED RESULTS

A new ENI to be created with 3 secondary private IPs.
##### ACTUAL RESULTS

The module finds some existing ENI and just returns "changed":"false". This breaks creating ENIs completely.

```
TASK [Create ENI for eth0 on Edge instances] ***********************************
ok: [localhost] => {"changed": false, "interface": {"attachment": {"attach_time": "2016-04-05T21:04:51.000Z", "attachment_id": "eni-attach-476e1abb", "delete_on_termination": false, "device_index": 1, "instance_id": "i-f3a7d668", "status": "attached"}, "description": "", "groups": {"sg-5f4c6f27": "GWAAS-SG-Iavor-Edge"}, "id": "eni-df31ee8f", "mac_address": "0e:e6:a9:ff:d4:4b", "owner_id": "953738397088", "private_ip_address": "172.16.2.151", "private_ip_addresses": [{"primary_address": true, "private_ip_address": "172.16.2.151"}, {"primary_address": false, "private_ip_address": "172.16.2.230"}, {"primary_address": false, "private_ip_address": "172.16.2.61"}, {"primary_address": false, "private_ip_address": "172.16.2.62"}], "source_dest_check": true, "status": "in-use", "subnet_id": "subnet-72870d2a", "vpc_id": "vpc-617a6b05"}}
```

Copied from original issue: ansible/ansible-modules-extras#2956

**Comment by ansibot** [comment]
From @KpaBap on 2016-09-18T23:05:31Z

The bug appears to be in the "find_eni" function - it's only using the subnet as a filter to find existing ENIs.

**Comment by ansibot** [comment]
From @ansibot on 2016-09-18T23:05:31Z

@wimnat, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @wimnat on 2016-09-18T23:05:31Z

Confirmed.

How would you like this solved?

@filipenf made changes to this module to make it more idempotent but it looks like maybe now it's gone too far.

Perhaps another parameter to say 'always create new'? 

needs_info

**Comment by ansibot** [comment]
From @filipenf on 2016-09-18T23:05:31Z

You're right @wimnat. What we could do would be to use the Name tag additionally to the ip/subnet combination. I think this would be cleaner than having an always_create and would be closer to how the ec2 module works for launching instances.

**Comment by ansibot** [comment]
From @ansibot on 2016-09-18T23:05:31Z

@KpaBap ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @hankok on 2016-09-18T23:05:31Z

I got the same issue, not fixed in the latest build Ansible 2.3 and 2.2

**Comment by ansibot** [comment]
From @ansibot on 2016-09-18T23:05:31Z

@KpaBap ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @Sodki on 2016-09-18T23:05:31Z

I'm affected by the same issue after updating to Ansible 2.2.

My use case is that I need to know the IP address before I launch a new machine. My solution was to create an ENI, obtain the IP address from it and then launch the machine, but my code is now returning an existing ENI instead of creating a new one. This is what I'm doing (and failing):

```
- ec2_eni:
    profile: "{{ profile }}"
    region: "{{ region }}"
    security_groups: "{{ security_group }}"
    subnet_id: "{{ subnet_id }}"
  register: new_eni
```

From the above, there is nothing that can identify a new ENI, so it shouldn't try to be idempotent. We need to figure out in which situations we want to be idempotent, otherwise this will cause all sorts of issues.

**Comment by ansibot** [comment]
From @Sodki on 2016-09-18T23:05:31Z

@wimnat Can you remove the `needs_info` label, please? @KpaBap might be unresponsive, but there are other responsive people affected by this issue (like me) and it would be a shame to see it closed by @ansibot.

Having a new parameter to always create an ENI is a workaround, but it still feels that the current idempotency logic is flawed.

**Comment by ansibot** [comment]
From @wimnat on 2016-09-18T23:05:31Z

So how about we just create new if IP isn't provided?

If the user wanted to NOT create a new eni if a subnet already had one then you could just use eni_facts module to count the number of existing ENIs and then use a when: statement on the eni module

**Comment by ansibot** [comment]
From @ansibot on 2016-09-18T23:05:31Z

@KpaBap ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @morganchristiansson on 2016-09-18T23:05:31Z

It's currently not possible to create a detached ENI with ec2_eni module.

Workaround is to not use ansible.

Duplicate issues: https://github.com/ansible/ansible/issues/19972 https://github.com/ansible/ansible/issues/19452

**Comment by ansibot** [comment]
From @morganchristiansson on 2016-09-18T23:05:31Z

I've come up with this workaround:
```yaml
- name: Allocate ENI
  # ec2_eni:
  #   subnet_id: subnet-e1588e8a
  #   region: "{{ aws_region }}"
  # register: eni
  shell: aws ec2 create-network-interface --subnet-id subnet-xxxxxxxx --region {{ aws_region }}
  register: output_raw

- set_fact:
    output_json: "{{ output_raw.stdout|from_json }}"
- set_fact:
    eni:
      interface:
        id: "{{ output_json.NetworkInterface.NetworkInterfaceId }}"
        private_ip_address: "{{ output_json.NetworkInterface.PrivateIpAddresses[0].PrivateIpAddress }}"
```

**Comment by ansibot** [comment]
cc @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@KpaBap This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@KpaBap You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
